common/lockmanager_test.go

```diff
@@ -0,0 +1,355 @@
+    // Set the timestamp for the lock to be considered stale.
+    val.(*lockData).timestamp = time.Now().Add(-1 * (staleDuration + (1 * time.Minute)));
+
+    threadExecutionBlock.Add(1);
+    timestampBlock.Add(1);
+
+    go func() {
+        defer threadExecutionBlock.Done();
+
+        // Wait for the lock's timestamp to be reset.
+        timestampBlock.Wait();
+
+        RemoveStaleLocksOnce();
+    }()
+
+    // Reset the locks timestamp.
+    Lock(key1);
+    Unlock(key1);
+    
+    // Let the goroutine continue its execution after resetting the timestamp.
+    timestampBlock.Done();
+
+    // Wait for the go routine to finish its execution.
+    threadExecutionBlock.Wait();
+    
+    // Check if the lock still exists in the lock map.
+    _, exists := lockMap.Load(key1);
+    // Test if the non stale lock got removed.
+    if (!exists) {
+        test.Fatalf("Failed to retain lock even though it was accessed concurrently.");
+    }
+} 
+
+func TestStaleLockWithLockedKey(test *testing.T) {
+    Clear();
+    defer Clear();
+
+    key1 := "testkey1";
+    staleDuration := time.Duration(config.STALELOCK_DURATION_SECS.Get()) * time.Second;
+    var staleCheckBlock sync.WaitGroup;
+
+    Lock(key1);
+
+    val, _ := lockMap.Load(key1); 
+    // Set the timestamp for a lock to be considered stale.
+    val.(*lockData).timestamp = time.Now().Add(-1 * (staleDuration + (1 * time.Second)));
+    
+    staleCheckBlock.Add(1);
+
+    // Start a goroutine to check if the lock is stale.
+    go func() {
+        defer staleCheckBlock.Done();
+
+        RemoveStaleLocksOnce();
+    }()
+    
+    // Wait for the goroutine to finish its execution.
+    staleCheckBlock.Wait();
+
+    // Check if the lock still exists in the lock map and return it.
+    _, exists := lockMap.Load(key1);
+    // Test if the Locked lock got removed.
+    if (!exists) {
+        test.Fatalf("Failed to retain lock even though it was locked.");
+    }
+}
+
+func TestStaleLockPassesFirstCheckButNotSecond(test *testing.T) {
+    Clear();
+    defer Clear();
+
+    key1 := "testkey1";
+    staleDuration := time.Duration(config.STALELOCK_DURATION_SECS.Get()) * time.Second;
+    var finishThreadBlock sync.WaitGroup;
+
+    Lock(key1);
+    Unlock(key1);
+
+    val, _ := lockMap.Load(key1);
+    lockData := val.(*lockData);
+    // Set the timestamp for a lock to be considered stale.
+    lockData.timestamp = time.Now().Add(-1 * (staleDuration + (1 * time.Minute)));
+
+    finishThreadBlock.Add(1);
+
+    // Lock the lockManagerMutex to to give the main thread time to reset the lock's timestamp.
+    lockManagerMutex.Lock();
+
+    // This thread will pass the first guard but wait until the lock's timestamp gets reset
+    // to pass the second guard.
+    go func() {
+        defer finishThreadBlock.Done();
+
+        RemoveStaleLocksOnce();
+    }()
+
+    // Small sleep to give the thread time to pass the first check in RemoveStaleLocksOnce().
```

Comment:
You need another wait group before this sleep.
You generally don't want to just use a sleep to wait for a thread to start.
There is a TON of variability on how fast a thread can start.
So, use a lock (wait group in this case) to wait for a thread to start, and then use a sleep to let it run.